File: lib/data/repositories/tour_repository.dart
import 'dart:convert';

import 'package:drift/drift.dart';

import '../../core/animal_counts_from_prestations.dart';
import '../../core/animal_counts_merge.dart';
import '../../core/tour_stop_prestations_normalizer.dart';
import '../../domain/models/coordinates.dart';
import '../../domain/models/tour.dart';
import '../../domain/models/tour_stop.dart';
import '../../domain/models/tour_stop_prestation.dart';
import '../../infra/db/app_database.dart';

class TourStopDraft {
  final int? clientId;
  final String clientNameSnapshot;
  final int orderIndex;
  final int estimatedArrivalMinutes;
  final int estimatedDepartureMinutes;
  final List<TourStopPrestation> plannedPrestations;
  final int feeShareCents;

  const TourStopDraft({
    required this.clientNameSnapshot,
    required this.orderIndex,
    required this.estimatedArrivalMinutes,
    required this.estimatedDepartureMinutes,
    required this.plannedPrestations,
    required this.feeShareCents,
    this.clientId,
  });
}

class TourDraft {
  final DateTime plannedDate;
  final int startTimeMinutes;
  final int totalDistanceMeters;
  final int totalDriveSeconds;
  final int totalTravelFeeCents;
  final String? notes;
  final List<TourStopDraft> stops;

  /// Polyline ORS du trajet complet (base → stops → base). Nullable :
  /// si l'appel ORS échoue (offline, quota), on persiste null et la map
  /// tombe en lignes droites côté UI.
  final List<Coordinates>? routeGeometry;

  const TourDraft({
    required this.plannedDate,
    required this.startTimeMinutes,
    required this.totalDistanceMeters,
    required this.totalDriveSeconds,
    required this.totalTravelFeeCents,
    required this.stops,
    this.notes,
    this.routeGeometry,
  });
}

/// Encode/decode helpers pour le champ JSON `tours.routeGeometry`.
/// Format compact : `[[lat, lon], [lat, lon], ...]`.
String? encodeRouteGeometry(List<Coordinates>? coords) {
  if (coords == null || coords.isEmpty) return null;
  return jsonEncode([
    for (final c in coords) [c.lat, c.lon],
  ]);
}

List<Coordinates>? decodeRouteGeometry(String? json) {
  if (json == null || json.isEmpty) return null;
  final raw = jsonDecode(json);
  if (raw is! List) return null;
  return [
    for (final p in raw)
      if (p is List && p.length >= 2)
        Coordinates(lat: (p[0] as num).toDouble(), lon: (p[1] as num).toDouble()),
  ];
}

class TourWithStops {
  final Tour tour;
  final List<TourStop> stops;
  const TourWithStops({required this.tour, required this.stops});
}

class TourRepository {
  final AppDatabase _db;
  TourRepository(this._db);

  Future<int> plan(TourDraft draft) async {
    return _db.transaction(() async {
      final now = DateTime.now().millisecondsSinceEpoch;
      final tourId = await _db.into(_db.toursTable).insert(
            ToursTableCompanion.insert(
              plannedDate: _toEpochDay(draft.plannedDate),
              startTimeMinutes: draft.startTimeMinutes,
              status: 'planned',
              totalDistanceMeters: draft.totalDistanceMeters,
              totalDriveSeconds: draft.totalDriveSeconds,
              totalTravelFeeCents: draft.totalTravelFeeCents,
              notes: Value(draft.notes),
              createdAt: now,
              routeGeometry: Value(encodeRouteGeometry(draft.routeGeometry)),
            ),
          );
      for (final s in draft.stops) {
        await _db.into(_db.tourStopsTable).insert(
              TourStopsTableCompanion.insert(
                tourId: tourId,
                clientId: Value(s.clientId),
                clientNameSnapshot: s.clientNameSnapshot,
                orderIndex: s.orderIndex,
                estimatedArrivalMinutes: s.estimatedArrivalMinutes,
                estimatedDepartureMinutes: s.estimatedDepartureMinutes,
                plannedPrestations: Value(
                    normalizeTourStopPrestations(s.plannedPrestations)),
                feeShareCents: s.feeShareCents,
              ),
            );
      }
      return tourId;
    });
  }

  Future<TourWithStops?> findById(int id) async {
    final tourRow = await (_db.select(_db.toursTable)
          ..where((t) => t.id.equals(id)))
        .getSingleOrNull();
    if (tourRow == null) return null;
    final stopRows = await (_db.select(_db.tourStopsTable)
          ..where((s) => s.tourId.equals(id))
          ..orderBy([(s) => OrderingTerm.asc(s.orderIndex)]))
        .get();
    return TourWithStops(
      tour: _tourFromRow(tourRow),
      stops: stopRows.map(_stopFromRow).toList(),
    );
  }

  Future<List<Tour>> listAll() async {
    final rows = await (_db.select(_db.toursTable)
          ..orderBy([(t) => OrderingTerm.desc(t.plannedDate)]))
        .get();
    return rows.map(_tourFromRow).toList();
  }

  /// Mark a tour as completed. The caller passes a map keyed by `stop.id`
  /// containing the actual prestations and an optional note. For each stop
  /// with a non-null `clientId`, per-category counts are derived from the
  /// actual prestations (MAX rule, libres ignored) and merged into the
  /// client's stored animals; `lastInterventionDate` is bumped to the tour's
  /// planned date.
  Future<void> markCompleted(
    int id,
    Map<int, ({List<TourStopPrestation> actuals, String? note})> actuals,
  ) async {
    await _db.transaction(() async {
      final tour = await (_db.select(_db.toursTable)
            ..where((t) => t.id.equals(id)))
          .getSingle();
      final now = DateTime.now().millisecondsSinceEpoch;
      final tourDateUtc = DateTime.fromMillisecondsSinceEpoch(
          tour.plannedDate * 86400000,
          isUtc: true);
      final tourDate =
          DateTime(tourDateUtc.year, tourDateUtc.month, tourDateUtc.day);

      // 1. Status flip.
      await (_db.update(_db.toursTable)..where((t) => t.id.equals(id))).write(
        ToursTableCompanion(
          status: const Value('completed'),
          completedAt: Value(now),
        ),
      );

      // 2. For each stop: persist actuals + note, then sync the linked client.
      final stopRows = await (_db.select(_db.tourStopsTable)
            ..where((s) => s.tourId.equals(id)))
          .get();

      for (final s in stopRows) {
        final entry = actuals[s.id];
        if (entry == null) continue;
        final normalized = normalizeTourStopPrestations(entry.actuals);

        await (_db.update(_db.tourStopsTable)
              ..where((t) => t.id.equals(s.id)))
            .write(
          TourStopsTableCompanion(
            actualPrestations: Value(normalized),
            interventionNote: Value(entry.note),
          ),
        );

        final cid = s.clientId;
        if (cid != null) {
          // Derive per-category counts from the actual prestations (MAX rule,
          // libres ignored) and merge them into the client's stored animals.
          final derived = animalCountsFromPrestations(normalized);
          final clientRow = await (_db.select(_db.clientsTable)
                ..where((c) => c.id.equals(cid)))
              .getSingleOrNull();
          if (clientRow != null) {
            final mergedAnimals = mergeAnimalCountsByCategory(
              clientRow.animals,
              derived,
            );
            await (_db.update(_db.clientsTable)
                  ..where((c) => c.id.equals(cid)))
                .write(
              ClientsTableCompanion(
                animals: Value(mergedAnimals),
                lastInterventionDate: Value(tourDate.millisecondsSinceEpoch),
                updatedAt: Value(now),
              ),
            );
          }
        }
      }
    });
  }

  /// Replace the contents of a planned tour: tour totals, date, start time,
  /// notes, and the entire stops list. Stop ids are not preserved (we delete
  /// then reinsert) — there is no consumer that holds onto them.
  ///
  /// Status, completedAt, id and createdAt are intentionally left untouched.
  /// The caller (UI) must only invoke this for tours with status `planned`.
  Future<void> update(int id, TourDraft draft) async {
    await _db.transaction(() async {
      await (_db.update(_db.toursTable)..where((t) => t.id.equals(id))).write(
        ToursTableCompanion(
          plannedDate: Value(_toEpochDay(draft.plannedDate)),
          startTimeMinutes: Value(draft.startTimeMinutes),
          totalDistanceMeters: Value(draft.totalDistanceMeters),
          totalDriveSeconds: Value(draft.totalDriveSeconds),
          totalTravelFeeCents: Value(draft.totalTravelFeeCents),
          notes: Value(draft.notes),
          routeGeometry: Value(encodeRouteGeometry(draft.routeGeometry)),
        ),
      );

      await (_db.delete(_db.tourStopsTable)
            ..where((s) => s.tourId.equals(id)))
          .go();

      for (final s in draft.stops) {
        await _db.into(_db.tourStopsTable).insert(
              TourStopsTableCompanion.insert(
                tourId: id,
                clientId: Value(s.clientId),
                clientNameSnapshot: s.clientNameSnapshot,
                orderIndex: s.orderIndex,
                estimatedArrivalMinutes: s.estimatedArrivalMinutes,
                estimatedDepartureMinutes: s.estimatedDepartureMinutes,
                plannedPrestations: Value(
                    normalizeTourStopPrestations(s.plannedPrestations)),
                feeShareCents: s.feeShareCents,
              ),
            );
      }
    });
  }

  Future<void> delete(int id) async {
    await (_db.delete(_db.toursTable)..where((t) => t.id.equals(id))).go();
  }

  int _toEpochDay(DateTime d) =>
      DateTime.utc(d.year, d.month, d.day).millisecondsSinceEpoch ~/
          86400000;

  Tour _tourFromRow(TourRow row) => Tour(
        id: row.id,
        plannedDate: () {
          final utc = DateTime.fromMillisecondsSinceEpoch(
              row.plannedDate * 86400000,
              isUtc: true);
          return DateTime(utc.year, utc.month, utc.day);
        }(),
        startTimeMinutes: row.startTimeMinutes,
        status: row.status == 'completed'
            ? TourStatus.completed
            : TourStatus.planned,
        totalDistanceMeters: row.totalDistanceMeters,
        totalDriveSeconds: row.totalDriveSeconds,
        totalTravelFeeCents: row.totalTravelFeeCents,
        notes: row.notes,
        completedAt: row.completedAt == null
            ? null
            : DateTime.fromMillisecondsSinceEpoch(row.completedAt!),
        createdAt: DateTime.fromMillisecondsSinceEpoch(row.createdAt),
        routeGeometry: decodeRouteGeometry(row.routeGeometry),
      );

  TourStop _stopFromRow(TourStopRow row) => TourStop(
        id: row.id,
        tourId: row.tourId,
        clientId: row.clientId,
        clientNameSnapshot: row.clientNameSnapshot,
        orderIndex: row.orderIndex,
        estimatedArrivalMinutes: row.estimatedArrivalMinutes,
        estimatedDepartureMinutes: row.estimatedDepartureMinutes,
        plannedPrestations: row.plannedPrestations,
        actualPrestations: row.actualPrestations,
        interventionNote: row.interventionNote,
        feeShareCents: row.feeShareCents,
      );
}

File: src/data/repositories/tour-repository.ts
import { asc, desc, eq } from 'drizzle-orm';
import { tours, tourStops } from '@/infra/db/schema';
import type { Db } from '@/infra/db/client';
import { Tour, type TourStatus } from '@/domain/models/tour';
import { TourStop } from '@/domain/models/tour-stop';
import { Payment } from '@/domain/models/payment';

interface TourRow {
  id: string;
  scheduledDate: string | null;
  departureTime: string | null;
  title: string | null;
  baseLat: number;
  baseLng: number;
  status: string;
  totalDistanceKm: number | null;
  totalDriveSeconds: number | null;
  totalMinutes: number | null;
  totalRevenueCents: number | null;
  totalAnimalsCount: number | null;
  routeGeometry: string | null;
  notes: string | null;
  completedAt: string | null;
  createdAt: string;
  updatedAt: string;
}

interface TourStopRow {
  id: string;
  tourId: string;
  clientId: string;
  clientNameSnapshot: string | null;
  ordering: number;
  arrivalMinutes: number | null;
  departureMinutes: number | null;
  estimatedMinutes: number | null;
  travelFeeCents: number | null;
  plannedServices: string;
  actualServices: string | null;
  notes: string | null;
  completedAt: string | null;
  paymentMethodId: string | null;
  paymentMethodLabelSnapshot: string | null;
  isPaid: number;
  paidAt: string | null;
}

function tourFromRow(r: TourRow): Tour {
  return Tour.parse(r);
}

function stopToRow(s: TourStop) {
  return {
    id: s.id,
    tourId: s.tourId,
    clientId: s.clientId,
    clientNameSnapshot: s.clientNameSnapshot,
    ordering: s.ordering,
    arrivalMinutes: s.arrivalMinutes,
    departureMinutes: s.departureMinutes,
    estimatedMinutes: s.estimatedMinutes,
    travelFeeCents: s.travelFeeCents,
    plannedServices: JSON.stringify(s.plannedServices),
    actualServices: s.actualServices === null ? null : JSON.stringify(s.actualServices),
    notes: s.notes,
    completedAt: s.completedAt,
    paymentMethodId: s.payment.methodId,
    paymentMethodLabelSnapshot: s.payment.methodLabelSnapshot,
    isPaid: s.payment.isPaid ? 1 : 0,
    paidAt: s.payment.paidAt,
  };
}

function stopFromRow(r: TourStopRow): TourStop {
  return TourStop.parse({
    id: r.id,
    tourId: r.tourId,
    clientId: r.clientId,
    clientNameSnapshot: r.clientNameSnapshot,
    ordering: r.ordering,
    arrivalMinutes: r.arrivalMinutes,
    departureMinutes: r.departureMinutes,
    estimatedMinutes: r.estimatedMinutes,
    travelFeeCents: r.travelFeeCents,
    plannedServices: JSON.parse(r.plannedServices),
    actualServices: r.actualServices === null ? null : JSON.parse(r.actualServices),
    notes: r.notes,
    completedAt: r.completedAt,
    payment: {
      methodId: r.paymentMethodId,
      methodLabelSnapshot: r.paymentMethodLabelSnapshot,
      isPaid: r.isPaid === 1,
      paidAt: r.paidAt,
    },
  });
}

export interface TourWithStops {
  tour: Tour;
  stops: TourStop[];
}

export class TourRepository {
  constructor(private readonly db: Db) {}

  async byId(id: string): Promise<TourWithStops | null> {
    const tRows = await this.db.select().from(tours).where(eq(tours.id, id));
    if (!tRows[0]) return null;
    const sRows = await this.db
      .select()
      .from(tourStops)
      .where(eq(tourStops.tourId, id))
      .orderBy(asc(tourStops.ordering));
    return {
      tour: tourFromRow(tRows[0] as TourRow),
      stops: sRows.map((r) => stopFromRow(r as TourStopRow)),
    };
  }

  async listAll(): Promise<TourWithStops[]> {
    const tRows = await this.db.select().from(tours);
    const result: TourWithStops[] = [];
    for (const tr of tRows) {
      const sRows = await this.db
        .select()
        .from(tourStops)
        .where(eq(tourStops.tourId, (tr as TourRow).id))
        .orderBy(asc(tourStops.ordering));
      result.push({
        tour: tourFromRow(tr as TourRow),
        stops: sRows.map((r) => stopFromRow(r as TourStopRow)),
      });
    }
    return result;
  }

  async listByStatus(status: TourStatus): Promise<TourWithStops[]> {
    const orderColumn = (() => {
      switch (status) {
        case 'draft':     return desc(tours.updatedAt);
        case 'planned':   return asc(tours.scheduledDate);
        case 'completed': return desc(tours.scheduledDate);
      }
    })();
    const tRows = await this.db
      .select()
      .from(tours)
      .where(eq(tours.status, status))
      .orderBy(orderColumn);

    const result: TourWithStops[] = [];
    for (const tr of tRows) {
      const sRows = await this.db
        .select()
        .from(tourStops)
        .where(eq(tourStops.tourId, (tr as TourRow).id))
        .orderBy(asc(tourStops.ordering));
      result.push({
        tour: tourFromRow(tr as TourRow),
        stops: sRows.map((r) => stopFromRow(r as TourStopRow)),
      });
    }
    return result;
  }

  async upsertTour(t: Tour, stops: TourStop[]): Promise<void> {
    await this.db.insert(tours).values(t).onConflictDoUpdate({ target: tours.id, set: t });
    await this.db.delete(tourStops).where(eq(tourStops.tourId, t.id));
    for (const s of stops) {
      await this.db.insert(tourStops).values(stopToRow(s));
    }
  }

  async deleteTour(id: string): Promise<void> {
    await this.db.delete(tours).where(eq(tours.id, id));
  }

  async markStopCompleted(stopId: string, completedAt: string): Promise<void> {
    await this.db.update(tourStops).set({ completedAt }).where(eq(tourStops.id, stopId));
  }

  async markStopPayment(stopId: string, payment: Payment): Promise<void> {
    await this.db.update(tourStops).set({
      paymentMethodId: payment.methodId,
      paymentMethodLabelSnapshot: payment.methodLabelSnapshot,
      isPaid: payment.isPaid ? 1 : 0,
      paidAt: payment.paidAt,
    }).where(eq(tourStops.id, stopId));
  }

  async completeWithBilan(
    tourId: string,
    perStopActuals: Map<string, import('@/domain/models/tour-stop-service').TourStopService[]>,
    perStopNotes: Map<string, string | null>,
    perStopPayments: Map<string, Payment>,
    perStopTravelFees: Map<string, number>,
    completedAt: string
  ): Promise<void> {
    const result = await this.byId(tourId);
    if (!result) throw new Error('Tour introuvable');
    const { tour, stops } = result;

    const updatedStops = stops.map((s) => ({
      ...s,
      actualServices: perStopActuals.get(s.id) ?? s.plannedServices,
      notes: perStopNotes.has(s.id) ? perStopNotes.get(s.id) ?? null : s.notes,
      payment: perStopPayments.get(s.id) ?? s.payment,
      travelFeeCents: perStopTravelFees.has(s.id) ? perStopTravelFees.get(s.id)! : s.travelFeeCents,
      completedAt,
    }));

    await this.upsertTour(
      { ...tour, status: 'completed', completedAt, updatedAt: completedAt },
      updatedStops
    );
  }
}
